crates/ra_hir/src/ty/tests.rs

```diff
@@ -0,0 +1,71 @@
+use std::fmt::Write;
+use std::path::{PathBuf};
+use std::sync::Once;
+
+use flexi_logger::Logger;
+
+use ra_db::{SyntaxDatabase};
+use ra_syntax::ast::{self, AstNode};
+use test_utils::{project_dir, dir_tests};
+
+use crate::{
+    source_binder,
+    mock::MockDatabase,
+};
+
+fn infer_file(content: &str) -> String {
+    let (db, _, file_id) = MockDatabase::with_single_file(content);
+    let source_file = db.source_file(file_id);
+    let mut acc = String::new();
+    for fn_def in source_file
+        .syntax()
+        .descendants()
+        .filter_map(ast::FnDef::cast)
+    {
+        let func = source_binder::function_from_source(&db, file_id, fn_def)
+            .unwrap()
+            .unwrap();
+        let inference_result = func.infer(&db).unwrap();
+        for (syntax_ptr, ty) in &inference_result.type_for {
+            let node = syntax_ptr.resolve(&source_file);
+            write!(
+                acc,
+                "{} '{}': {}\n",
+                syntax_ptr.range(),
+                ellipsize(node.text().to_string().replace("\n", " "), 15),
+                ty
+            )
+            .unwrap();
+        }
+    }
+    acc
+}
+
+fn ellipsize(mut text: String, max_len: usize) -> String {
+    if text.len() <= max_len {
+        return text;
+    }
+    let ellipsis = "...";
+    let e_len = ellipsis.len();
+    let mut prefix_len = (max_len - e_len) / 2;
+    while !text.is_char_boundary(prefix_len) {
+        prefix_len += 1;
+    }
+    let mut suffix_len = max_len - e_len - prefix_len;
+    while !text.is_char_boundary(text.len() - suffix_len) {
+        suffix_len += 1;
+    }
+    text.replace_range(prefix_len..text.len() - suffix_len, ellipsis);
+    text
+}
+
+#[test]
+pub fn infer_tests() {
+    static INIT: Once = Once::new();
+    INIT.call_once(|| Logger::with_env().start().unwrap());
+    dir_tests(&test_data_dir(), &["."], |text, _path| infer_file(text));
+}
```

Comment:
I very much love such "data-driven" tests, where you just thrown two text-files at the thing and compare results.

However, for debugging, it's very useful to have a separate test for each dataset, so that one can easily launch just a single test and get useful logging, etc. We already [have that feature](https://youtu.be/Q2ced-S9Ip8?t=160) :) 

So, let's rewrite this as a bunch of functions like

```
#[test]
fn infers_types_for_primitives() {
    check_inference(r#"
fn test(a: u32, b: isize, c: !, d: &str) {
     a;
     b;
     c;
     d;
     1usize;
     1isize;
     "test";
     1.0f32;
 }
"#, r"
 [33; 34) 'd': [unknown]
 [88; 94) '1isize': [unknown]
 [48; 49) 'a': u32
 [55; 56) 'b': isize
 [112; 118) '1.0f32': [unknown]
 [76; 82) '1usize': [unknown]
 [9; 10) 'a': u32
 [27; 28) 'c': !
 [62; 63) 'c': !
 [17; 18) 'b': isize
 [100; 106) '"test"': [unknown]
 [42; 121) '{     ...f32; }': ()
 [69; 70) 'd': [unknown]
")
}
```

Reply:
True, though I like the ease of just deleting the file and rerunning the test to regenerate it, but admittedly it's just a matter of copy/pasting the expected result. Ideally, we would both be able to run single tests and to automatically update them...